cirq/testing/deprecation.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from contextlib import contextmanager
+
+ALLOW_DEPRECATION_IN_TEST = 'ALLOW_DEPRECATION_IN_TEST'
+
+
+@contextmanager
+def allow_deprecation():
+    """Allows deprecated functions, classes, decorators in tests.
+
+    It acts as a contextmanager that can be used in with statements:
+    >>> with allow_deprecation():
+    >>>     # do something deprecated
+    """
+    os.environ[ALLOW_DEPRECATION_IN_TEST] = 'True'
```

Comment:
Why are you doing this via environment variables? I thought originally you wanted to leave open the possibility that people running pytest could set this environment variable themselves to run things the old way, i.e. deprecation warnings are warnings. However, this function will actually unset the environment variable any time the context manager is used, so maybe that wasn't the intent. 

Have you considered using a global variable in `cirq/_compat.py` that you can set/unset instead? I think that would be a little more hygienic

Reply:
I want to keep allowing the deprecation to very narrow, specific cases: when we know that there is a deprecated function and we want to test that behavior. Any other time, things should fail while running in pytest. 

Env var vs global var - I considered that, it is essentially the same, but at least there is a symmetry with the pytest method. I have no strong opinion there - why do you think the hygiene is better with a global var compared to an env var?

Reply:
What symmetry with what pytest method? 

Hygiene -- it blows away any user-set value for `ALLOW_DEPRECATION_IN_TEST` which the user may be using for another process, in their bashrc, etc. It's the most global of all global variables :)